src/transformers/models/wav2vec2/tokenization_wav2vec2.py

```diff
@@ -324,28 +326,33 @@ def _get_word_offsets(
         offsets: Dict[str, Union[str, float]], word_delimiter_char: str = " "
     ) -> Dict[str, Union[str, float]]:
         word_offsets = []
-        final_offset_idx = len(offsets) - 1
 
+        last_state = "SPACE"
+        word = ""
+        start_offset = 0
+        end_offset = 0
         for i, offset in enumerate(offsets):
-            # define previous, next and current char
             char = offset["char"]
-            prev_char = offsets[i - 1]["char"] if i > 0 else None
-            next_char = offsets[i + 1]["char"] if i < final_offset_idx else None
-
-            # derive whether word begins, ends and whether current char is in word
-            word_begin = (i == 0 and char != word_delimiter_char) or (prev_char == word_delimiter_char)
-            word_end = (i == final_offset_idx and char != word_delimiter_char) or (next_char == word_delimiter_char)
-            char_is_in_word = char != word_delimiter_char
-
-            if word_begin:
-                word_offset = {"word": "", "start_offset": offset["start_offset"]}
-
-            if word_end:
-                word_offset["end_offset"] = offset["end_offset"]
-                word_offsets.append(word_offset)
-
-            if char_is_in_word:
-                word_offset["word"] += offset["char"]
+            state = "SPACE" if char == word_delimiter_char else "WORD"
```

Comment:
I don't really understand why we have to change the code in `tokenization_wav2vec2.py` if the code in the pipeline is not aligned. We should only change the pipeline code IMO no?

Reply:
Like I don't understand what was wrong here previously

Reply:
No there was a bug:

> Have to change the tokenization word state machine. With chunking we can have 2 chunks
being concatenated, both containing a separate space. Hopefully this code is readable enough.

The core thing is that the end `tokens` I could present the `decode` function would be

[ ....., 0, 0, 0, 4, 4, 4, 0, 0, 0]  + [ 0, 0, 4, 0, 0, 0]

That created `char_offsets`  with 2 consecutive {char: " ", ...}, (that's ok and really what are in the tokens). and the current code for words would output *twice* the previously seen word (here `{"EXISTS, ..}, {"EXIST, ...}` since it assumed that a space meant "new word" all the time.

Making the modification as a full rewrite makes the final code simpler IMO.

Reply:
Ok good for me!

Reply:
New function looks clean to me